integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,87 @@
+package kotlinx.coroutines.experimental.androidx.lifecycle
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event.*
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.experimental.CoroutineScope
+import kotlinx.coroutines.experimental.Dispatchers
+import kotlinx.coroutines.experimental.Job
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Creates a [Job] that will be cancelled when the [Lifecycle] encounters [cancelEvent]
+ * (which defaults to [ON_DESTROY]).
+ *
+ * Note that [cancelEvent] supports only "**down**" events. That means that only [ON_PAUSE],
+ * [ON_STOP] and [ON_DESTROY] are supported.
+ *
+ * If the [Lifecycle] is already destroyed, then the returned [Job] comes already cancelled.
+ */
+fun Lifecycle.createJob(cancelEvent: Lifecycle.Event = ON_DESTROY): Job {
+    require(cancelEvent in allowedCancelEvents) { "$cancelEvent is forbidden for createJob(…)." }
+    return Job().also { job ->
+        if (currentState == Lifecycle.State.DESTROYED) job.cancel()
```

Comment:
Unless done in the main thread, this check is not thread safe, as the currentState can change and the stateChanged event can happen between lines 24 and 25 without being captured.

Reply:
You're commenting on outdated code, please use the "Files changed" tab instead of a specific commit.
However, the logic you're pinpointing has not changed so much. This is just a fast path. If it goes to the `else` case, then, the observer is added, and it'll receive the events, even if it's already destroyed. I just added `Test observer is called after destroy` to check this.

Reply:
In a few words: The way AndroidX lifecycles are implemented makes this thread safe.

Reply:
Okay, I didn't know the latest state would fire after subscribing. LGTM then.